types/wlr_surface.c

```diff
@@ -301,29 +301,8 @@ static void surface_apply_damage(struct wlr_surface *surface) {
 	}
 
 	if (surface->buffer != NULL && surface->buffer->released) {
-		pixman_region32_t damage;
-		pixman_region32_init(&damage);
-		pixman_region32_copy(&damage, &surface->current.buffer_damage);
-
-		pixman_region32_t surface_damage;
-		pixman_region32_init(&surface_damage);
-		pixman_region32_copy(&surface_damage, &surface->current.surface_damage);
-		wlr_region_transform(&surface_damage, &surface_damage,
-			wlr_output_transform_invert(surface->current.transform),
-			surface->current.width, surface->current.height);
-		wlr_region_scale(&surface_damage, &surface_damage,
-			surface->current.scale);
-		pixman_region32_union(&damage, &damage, &surface_damage);
-		pixman_region32_fini(&surface_damage);
-
-		pixman_region32_intersect_rect(&damage, &damage, 0, 0,
-			surface->current.buffer_width, surface->current.buffer_height);
-
 		struct wlr_buffer *updated_buffer =
-			wlr_buffer_apply_damage(surface->buffer, resource, &damage);
-
-		pixman_region32_fini(&damage);
-
+			wlr_buffer_apply_damage(surface->buffer, resource, &surface->buffer_damage);
```

Comment:
This doesn't work, because in case the surface moves (`sx` and `sy` are changed), `surface->buffer_damage` is not empty (contains the previous and the new buffer rectangles). However in this case we don't need to re-upload the texture.

Reply:
Hm, so there needs to a way to keep track of *why* the buffer was damaged, maybe? Or accumulating move damage separately. See my comment about removing move/resize damage from this stage in the pipeline as well.

Reply:
Nah, the buffer could have been moved _and_ damaged. We just need to use `surface->current.{buffer,surface}_damage`.

Originally the `buffer_damage` included these move/resize damaged regions as well because there wasn't `surface->previous`. Now that we have this, maybe we could make `buffer_damage` only contain the "real" damage as you'd like, but also provide a new utility function to compute the damage including move/resize. What do you think?

Reply:
I think a separation does make sense: a client `buffer_damage` (used in `surface_apply_damage`) plus a `total_damage`/`output_damage`/something? We could set them independently in `surface_update_damage` and take the union later, or treat the second variable as the union anyway.

Or, we could accumulate the total damage (resize/move) in terms of surface coordinates, which makes output damage calculation easier (no need to transform again).

Reply:
>I think a separation does make sense: a client buffer_damage (used in surface_apply_damage) plus a total_damage/output_damage/something?

I was just thinking of exposing the client damage in `surface->buffer_damage` and adding a new function (`wlr_surface_effective_damage`?) that would compute the total damage.

>Or, we could accumulate the total damage (resize/move) in terms of surface coordinates, which makes output damage calculation easier (no need to transform again).

I decided not to do so so that if you have a surface with scale=2 and an output with scale=2 you don't have to divide by 2 and then multiply by 2 the region, you can just keep it as-is. The Wayland protocol kind of encourages people to use buffer damage instead of surface damage (aka it would've been better if we only had buffer damage).

Reply:
> I was just thinking of exposing the client damage in surface->buffer_damage and adding a new function (wlr_surface_effective_damage?) that would compute the total damage.

I agree with this approach, but where would you call it? If it's during output damage calculation, it would presumably break sway. Maybe after calling `surface_apply_damage` in `surface_commit_pending`?

> The Wayland protocol kind of encourages people to use buffer damage instead of surface damage

I agree; I was just recommending surface damage for resize/move damages, not client damage.

(I'm sitting on IRC if you'd rather discuss there.)

Reply:
>If it's during output damage calculation, it would presumably break sway.

We can break this API and update sway. I'd just let compositors call it if they need it.

Reply:
If that's the case, +1 for `wlr_surface_(get/calculate_?)effective_damage`. Presumably, one: it'll handle resize/move damage; two: handle buffer damage to surface damage transformation; and three: return in terms of surface coordinates?

Reply:
I'd like to keep buffer coords. Compositors need to convert to output coords anyway, so doing buffer->surface in addition to surface->output is little work. Here's how it looks:

https://github.com/swaywm/wlroots/blob/936ec022664625f70fbcb22de6794d774cca159f/rootston/output.c#L708

Reply:
I'm not a very big fan of the idea of dealing with move/resize damage in terms of buffer coords. Two points I've already said before: one, they'd go surface->buffer->surface->output; two, semantically it's not buffer damage. Finally: if/when viewports are implemented, the conversions would get messy pretty quickly.

(Sorry, busy weekend)

Reply:
Eh, I see. Yeah, that makes sense, even if it makes me a little bit sad inside.

So do we agree on:
* Keep state (`pending` and `current`) as is
* Make `buffer_damage` the sum of surface damage and buffer damage submitted by the client and clipped to the surface bounds (used when uploading textures and by compositors not using the helper function)
* Add a new `wlr_surface_get_effective_damage` helper function that handles move/resize and is in surface-local coords

Reply:
We could also handle resize in `buffer_damage`, and handle move in `wlr_surface_get_effective_damage`. Not sure what's best. This could make sense since the whole texture is re-uploaded when resized.

Reply:
Ack on everything.

I think it might make sense to split resize damage: `surface_update_damage` damages the whole buffer (in buffer coords) if required (which preserves surface bounds, and integrates with viewports easily); while `wlr_surface_get_effective_damage` gives the surface damage to the compositor, including `previous` surface dimensions.

Reply:
Maybe #277 is relevant here: the function could take care of opaque regions implicitly?

Reply:
>I think it might make sense to split resize damage

+1

>Maybe #277 is relevant here: the function could take care of opaque regions implicitly?

Hmm, #277 is a lot more complicated than this, and requires knowledge about other surfaces. We'd basically need to walk down the surfaces stack to collect the opaque region, and then walk up to render according to damage + opaque. It should be implemented in another standalone PR.

Reply:
Hm, alright. I'll add this function as soon as possible then. I'm assuming you're okay with all the current changes in the PR?

Reply:
Yeah, I think so, they make sense. I'll do another final review to make sure I've not missed anything after you send a new version.